lib/elixir/lib/module/types/pattern.ex

```diff
@@ -214,71 +217,88 @@ defmodule Module.Types.Pattern do
   # we add support for intersection types for the guard functions
   # -integer / +integer
-  def of_guard({{:., _, [:erlang, guard]}, _, [integer]}, _stack, context)
+  def of_guard({{:., _, [:erlang, guard]}, _, [integer]}, _expected, _stack, context)
       when guard in [:+, :-] and is_integer(integer) do
     {:ok, :integer, context}
   end
 
   # -float / +float
-  def of_guard({{:., _, [:erlang, guard]}, _, [float]}, _stack, context)
+  def of_guard({{:., _, [:erlang, guard]}, _, [float]}, _expected, _stack, context)
       when guard in [:+, :-] and is_float(float) do
     {:ok, :float, context}
   end
 
   # fun(args)
-  def of_guard({{:., _, [:erlang, guard]}, _, args} = expr, stack, context) do
-    stack = push_expr_stack(expr, stack)
-    {param_types, return_type} = guard_signature(guard, length(args))
+  def of_guard({{:., _, [:erlang, guard]}, _, args} = expr, expected, stack, context) do
     type_guard? = type_guard?(guard)
     {consider_type_guards?, keep_guarded?} = stack.type_guards
+    # TODO: Handle no matching signature
+    signature = guard_signature(guard, length(args))
 
     # Only check type guards in the context of and/or/not,
     # a type guard in the context of is_tuple(x) > :foo
     # should not affect the inference of x
     if not type_guard? or consider_type_guards? do
+      stack = push_expr_stack(expr, stack)
+      param_unions = signature_to_param_unions(signature, context)
       arg_stack = %{stack | type_guards: {false, keep_guarded?}}
+      mfa = {:erlang, guard, length(args)}
 
       with {:ok, arg_types, context} <-
-             map_reduce_ok(args, context, &of_guard(&1, arg_stack, &2)),
-           {:ok, context} <- unify_call(arg_types, param_types, stack, context) do
-        {arg_types, guard_sources} =
-          case arg_types do
-            [{:var, index} | rest_arg_types] when type_guard? ->
-              guard_sources = Map.put_new(context.guard_sources, index, :guarded)
-              {rest_arg_types, guard_sources}
-
-            _ ->
-              {arg_types, context.guard_sources}
-          end
-
-        guard_sources =
-          Enum.reduce(arg_types, guard_sources, fn
-            {:var, index}, guard_sources ->
-              Map.update(guard_sources, index, :fail, &guarded_if_keep_guarded(&1, keep_guarded?))
-
-            _, guard_sources ->
-              guard_sources
-          end)
-
+             map_reduce_ok(Enum.zip(args, param_unions), context, fn {arg, param}, context ->
+               of_guard(arg, param, arg_stack, context)
+             end),
+           {:ok, return_type, context} <-
+             unify_call(arg_types, signature, expected, mfa, stack, context, type_guard?) do
+        guard_sources = guard_sources(arg_types, type_guard?, keep_guarded?, context)
         {:ok, return_type, %{context | guard_sources: guard_sources}}
       end
     else
+      # Assume type guard
```

Comment:
Instead of assuming, should we define two distinct maps? Type guards will have to be specially handled throughout the source code anyway.

Reply:
The assumption is that type guards only have one clause so the match below should always succeed. Type guards always return booleans so we can remove the match completely and just return `:boolean`. WDYT?